Fix export_enum does not work with String variable.

strank opened the pull request:
Fixes #54828, should accept what is mentioned in the documentation now.

Splits export_annotations into export_annotation (for @export and without template parameters) and typed_export_annotations (for @export_xxx). The latter adds two more template parameters to accept more than the expected variable type: either one that can be converted (like float for int), or one that replaces the default one (like String for int when using export_enum).

The two new template parameters have defaults, so they do not need to be provided. And maybe they should really be lists of types, so there can be several convertable and alternative types not just one? E.g. export_enum could also accept StringName I think? And maybe allowing float instead of int is not always wanted for all export_xxx that accept int?

<!--
Pull requests should always be made for the `master` branch first, as that's
where development happens and the source of all future stable release branches.

Relevant fixes are cherry-picked for stable branches as needed.

Do not create a pull request for stable branches unless the change is already
available in the `master` branch and it cannot be easily cherry-picked.
Alternatively, if the change is only relevant for that branch (e.g. rendering
fixes for the 3.2 branch).
-->

strank committed:
Fix export_enum does not work with String variable.

Fixes #54828

akien-mga commented:
@strank Just FYI (not related to this PR), we just created a #gdscript channel on https://chat.godotengine.org if you're interested in joining dev discussions with other contributors interested in GDScript :)

Chaosus commented:
@strank This PR needs to be rebased

pierpo commented:
Hey! Any news for this PR? What are the next steps? Should we help in any way? :blush:

AttackButton commented:
Interestingly, https://github.com/Yuminous/godot/pull/19 says it fixed the same bug, but that is not the case.

strank commented:
Rebased and updated, should merge without conflicts again.

> Hey! Any news for this PR? What are the next steps? Should we help in any way? 😊

Thanks @pierpo :) but I think this is now just waiting for a core dev to get to it.

anvilfolk commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -3731,87 +3825,21 @@ bool GDScriptParser::export_annotations(const AnnotationNode *p_annotation, Node
+		if (export_type.builtin_type == t_alternative_type) {
+			variable->export_info.type = t_alternative_type;
+		} else if (export_type.builtin_type == t_convertable_type) {
+			// convertable type (usually int <-> float) is ok
```
Is this incomplete or can this else if be removed and replaced by a comment elsewhere?

anvilfolk commented:
Needs another rebase and I'm not an expert but this all looks pretty sensible to me. I'll be happy to compile and test locally once it's been rebased!

anvilfolk commented:
> The latter adds two more template parameters to accept more than the expected variable type: either one that can be converted (like float for int), or one that replaces the default one (like String for int when using export_enum).

Actually, is this necessary? We should be able to check whether types are compatible automatically and possibly convert them directly. I did something like this in #69416, it might be worth checking out. It might simplify the template a little? What do you think?

dalexeev commented:
I suggest closing this PR as the related issue has been fixed. However, thanks for your efforts!
